generate/generate.go

```diff
@@ -135,150 +163,300 @@ func Load(cfg *config.Tree) ([]LoadResult, error) {
+		log.Info().
+			Stringer("stack", stack.Path()).
+			Str("file", filename).
+			Msg("deleted file")
+		report.addDeletedFile(filename)
+	}
 
-			filename := file.Label()
-			path := filepath.Join(stackpath, filename)
-			logger := logger.With().
-				Str("filename", filename).
-				Logger()
+	logger.Debug().Msg("finished generating files")
+	return report
+}
 
-			if !file.Condition() {
-				logger.Debug().Msg("condition is false, ignoring file")
-				continue
-			}
+func doDirGeneration(cfg *config.Tree, evalctx *eval.Context) Report {
+	logger := log.With().
+		Str("action", "generate.doDirGeneration()").
+		Str("dir", cfg.Dir()).
+		Logger()
 
-			err := writeGeneratedCode(path, file)
-			if err != nil {
-				return failureReport(
-					report,
-					errors.E(err, "saving file %q", filename),
-				)
+	report := Report{}
+	var files []GenFile
+	blocks := cfg.Node.Generate.Files
+	if len(blocks) == 0 {
+		return report
+	}
+
+	for _, block := range blocks {
+		if block.Context != genfile.RootContext {
+			continue
+		}
+
+		if block.Label[0] != '/' {
+			targetDir := path.Dir(block.Label)
+			report.addFailure(project.NewPath("/"+targetDir), errors.E(
+				ErrInvalidGenBlockLabel,
+				"generate_file.context=root requires an absolute path but given %s",
+				block.Label))
+
+			return report
+		}
+
+		targetDir := project.NewPath(path.Dir(block.Label))
+
+		// root generate
+		file, err := genfile.Eval(block, evalctx)
+		if err != nil {
+			report.addFailure(targetDir, err)
+			return report
+		}
+		files = append(files, file)
+	}
+
+	errsmap := checkFileConflict(files)
+	if len(errsmap) > 0 {
+		if len(errsmap) > 0 {
+			for file, err := range errsmap {
+				targetDir := path.Dir(file)
+				report.addFailure(project.NewPath(targetDir), err)
 			}
+			return report
+		}
+	}
 
-			// Change detection + remove entries that got re-generated
-			removedFileBody, ok := removedFiles[filename]
-			if !ok {
-				log.Info().
-					Stringer("stack", stack.Path()).
-					Str("file", filename).
-					Msg("created file")
+	errsmap = validateRootGeneratedFiles(cfg, files)
+	if len(errsmap) > 0 {
+		for file, err := range errsmap {
+			targetDir := path.Dir(file)
+			report.addFailure(project.NewPath(targetDir), err)
+		}
+		return report
+	}
 
-				report.addCreatedFile(filename)
-			} else {
-				body := file.Header() + file.Body()
-				if body != removedFileBody {
+	removedFiles, errsmap := removeRootGeneratedFiles(cfg, files)
+	if len(errsmap) > 0 {
+		for file, err := range errsmap {
+			targetDir := path.Dir(file)
+			report.addFailure(project.NewPath(targetDir), err)
+		}
+		return report
+	}
+
+	logger.Debug().Msg("saving generated files")
+
+	for _, file := range files {
+		dirReport := dirReport{}
+		target := filepath.Join(cfg.RootDir(), file.Label())
+		basedir := project.NewPath(path.Dir(file.Label()))
+		filename := filepath.Base(file.Label())
+		logger := logger.With().
+			Str("filename", target).
+			Logger()
+
+		if !file.Condition() {
```

Comment:
Will this work if you have one generate_file with condition = false and another one with condition = true ? On the stack generation I remove all files, manipulate the map and report deleted files only in the end because of the multiple conditions case. It seems that here if I have one condition = true and one = false I will always report it was deleted (maybe even report the same file was deleted and then created ?).

Also multiple blocks with the same label are allowed when they have condition = false, and it seems this code would add the file multiple times in this case.

But I could be missing something.

Reply:
there's a `checkFileConflict()` function called before this that checks if only one target has condition=true, otherwise it fails. The at this point it just saves again the files that do need to be created.

Reply:
hmm I'm not so sure this is related to file conflicts or even saving/deleting files, I'm thinking on the report itself, but Ill check the code again and will try to explain better my concerns (could be missing something here).

Reply:
I get it now, sorry for the noise =P